mlx/backend/metal/reduce.cpp

```diff
@@ -36,37 +41,88 @@ void all_reduce_dispatch(
     MTL::ComputeCommandEncoder* compute_encoder,
     metal::Device& d) {
   // Get kernel and encode buffers
-  size_t in_size = in.size();
-  auto kernel = d.get_kernel("all_reduce_" + op_name + type_to_name(in));
 
-  compute_encoder->setComputePipelineState(kernel);
-  set_array_buffer(compute_encoder, in, 0);
-  set_array_buffer(compute_encoder, out, 1);
-  compute_encoder->setBytes(&in_size, sizeof(size_t), 2);
+  Dtype out_dtype = out.dtype();
+  bool is_out_64b_int = is_64b_int(out_dtype);
+  std::string kernel_name =
+      (is_out_64b_int) ? "all_reduce_no_atomics_" : "all_reduce_";
+  kernel_name += (op_name + type_to_name(in));
 
-  // Set grid dimensions
+  auto kernel = d.get_kernel(kernel_name);
 
   // We make sure each thread has enough to do by making it read in
   // at least n_reads inputs
   int n_reads = REDUCE_N_READS;
 
+  size_t in_size = in.size();
+  size_t intermediate_size;
+
   // mod_in_size gives us the groups of n_reads needed to go over the entire
   // input
   uint mod_in_size = (in_size + n_reads - 1) / n_reads;
-
   NS::UInteger thread_group_size = kernel->maxTotalThreadsPerThreadgroup();
   thread_group_size =
       mod_in_size > thread_group_size ? thread_group_size : mod_in_size;
+  uint simd_size = kernel->threadExecutionWidth();
+  thread_group_size =
+      ((thread_group_size + simd_size - 1) / simd_size) * simd_size;
 
   // If the number of thread groups needed exceeds 1024, we reuse threads groups
   uint n_thread_groups = safe_div(mod_in_size, thread_group_size);
   n_thread_groups = std::min(n_thread_groups, 1024u);
   uint nthreads = n_thread_groups * thread_group_size;
 
+  // Allocate intermediate array to store partial reduction results from all
+  // thread groups
+  array* output = &out;
```

Comment:
No need for that or the pointer to the intermediate array. See https://github.com/ml-explore/mlx/blob/65d0b8df9fb02a30445fc5431aa2d8e73ca0af2f/mlx/backend/metal/matmul.cpp#L251 for how to allocate an intermediate array.

Reply:
I did that so we conditionally create an intermediate array, since we don't need it for the path that uses atomics.